src/simulation_strategies/GeneralizedCliffordsSimulationStrategyV2.py

```diff
@@ -0,0 +1,176 @@
+        return substates
+
+    def _fill_r_sample(self) -> None:
+        self.r_sample = []
+        self.current_key = tuple(self.r_sample)
+        self.current_sample_probability = 1
+
+        while self.number_of_input_photons > len(self.r_sample):
+            if self.current_key not in self.pmfs:
+                self._calculate_new_layer_of_pmfs()
+            self._sample_from_latest_pmf()
+
+    def _calculate_new_layer_of_pmfs(self) -> None:
+        number_of_particle_to_sample = len(self.r_sample) + 1
+        possible_input_states = self._labeled_states[number_of_particle_to_sample]
+        corresponding_k_vectors = [[self.input_state[i] - state[i] for i in range(len(state))]
+                                   for state in possible_input_states]
+        weights = self._calculate_weights_from_k_vectors(array(corresponding_k_vectors, dtype=float))
+        # weights /= sum(weights)
+        self.possible_outputs[self.current_key] = self._generate_possible_output_states()
+
+        pmf = []
+
+        for output in self.possible_outputs[self.current_key]:
+            output = particle_state_to_modes_state(output, len(self.input_state))
+            pmf.append(0)
+            for i in range(len(possible_input_states)):
+                probability = self._calculate_outputs_probability(possible_input_states[i], output)
+                probability *= weights[i]
+                pmf[-1] += probability
+
+        self.pmfs[self.current_key] = pmf
+
+    def _calculate_weights_from_k_vectors(self, corresponding_k_vectors: ndarray) -> ndarray:
+        return array([self._calculate_multinomial_coefficient(vector)
+                      for vector in corresponding_k_vectors], dtype=float64)
+
+    @staticmethod
+    def _calculate_multinomial_coefficient(vector: ndarray) -> int:
+        """
+            Calculates multinomial coefficient of the vector, as proposed in Oszmaniec, Brod 2018
+            (above formula 39).
+            :param vector: Vector of which multinomial coefficient will be calculated.
+            :return: Multinomial coefficient of given vector.
+        """
+        particles_number = sum(vector)
+        multinomial_coefficient = factorial(particles_number)
+        for value in vector:
+            multinomial_coefficient /= factorial(value)
+
+        return multinomial_coefficient
+
+    def _generate_possible_output_states(self) -> List[ndarray]:
+        possible_output_states = []
+
+        for i in range(len(self.interferometer_matrix)):
+            new_possible_output = copy(self.r_sample)
+            new_possible_output = append(new_possible_output, [i])
+            possible_output_states.append(array(new_possible_output, dtype=int64))
+
+        return possible_output_states
+
+    def _calculate_outputs_probability(self, input_state: ndarray, output_state: ndarray) -> float:
+        permanent_calculator = ChinHuhPermanentCalculator(self.interferometer_matrix, input_state=input_state,
```

Comment:
Have you considered injecting a `type` that would be instantiated? This would allow creation of the wanted version, e.g. the parallel one.

Reply:
I'm on it.